core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1057,12 +1054,19 @@ public ShardPath shardPath() {
         return path;
     }
 
-    public void recoverFromStore(ShardRouting shard, StoreRecoveryService.RecoveryListener recoveryListener) {
+    public boolean recoverFromStore(ShardRouting shard) {
         // we are the first primary, recover from the gateway
         // if its post api allocation, the index should exists
         assert shard.primary() : "recover from store only makes sense if the shard is a primary shard";
         final boolean shouldExist = shard.allocatedPostIndexCreate();
-        storeRecoveryService.recover(this, shouldExist, recoveryListener);
+        StoreRecovery storeRecovery = new StoreRecovery(shardId, logger);
+        return storeRecovery.recoverFromStore(this, shouldExist, localNode);
+    }
+
+    public boolean restoreFromRepository(ShardRouting shard, IndexShardRepository repository) {
```

Comment:
There's a lot of boolean return values with no indication for what they mean, from looking at the code it looks like `true` means the recovery succeeded and `false` means it failed, maybe put it in javadoc somewhere? (there are a lot of places)